### ISSUE TITLE
`RuntimeWarning: invalid value encountered in cast  x_sample = x_sample.astype(np.uint8)`

### ISSUE THREAD

**Issue by realisticdreamer114514** [DenOfEquity/forge2_flux_kontext#12]
Here's a common ae-related bug, `RuntimeWarning: Invalid Value encountered in cast x_sample = x_sample.astype(np.uint8)`
I am using turbo flux lora from alimama, it still worked fine yesterday but now the preview starts out as usual but turns black right at step 3/4 (of 8~10). 
diffusers = 0.32.2, 
kontext ckpt is bnb nf4 safetensors with stock flux ae, stock clip l & t5xxl fp16 q4_0 gguf, diffusion in low bits bnb-nf4, input image with standard ratio & resized to bfl recommended. I use taesd for everything, might be it but what else do I need to check for diagnosing wheres the issue?

**Comment by realisticdreamer114514** [comment]
Switching to full vae encoding /decoding the previews are fine but only the final image is black

**Comment by DenOfEquity** [comment]
This isn't an extension issue.

There have been similar issues with many of the UIs, for a very long time. Sometimes corrupt model files, sometimes badly made LoRAs, often not enough precision, often unclear why - could be Torch/drivers/xformers differences between systems; impossible to track down unless experiencing the issue.
The place where the problem is reported is not the place where the error happens. AFAIK, the error happens earlier, either during VAE or inference, resulting in NaNs in the latent image. So doing a nan_to_num fix then clamping to [0,0, 1.0] before multiplication by 255 and conversion to numpy uint8 will avoid the warning but results can still be broken. It's also possible that fixing NaNs does solve it. Something like this, inserted before line 1039:  `x_sample = torch.clamp(torch.nan_to_num(x_sample, nan=0.0, posinf=1.0, neginf=0.0), min=0.0, max=1.0)`.

Your link refers to a different case - the rembg code is scaling the result to use the full range available, but not checking for potential division by zero.

**Comment by realisticdreamer114514** [comment]
Inserting the line you provided indeed removed the warning, but doesn't fix the bug at all. I tested with q3_k_s gguf and using no loras, still there 🤨. Has been testing in t2i mode only so far

btw I'm pretty much aware it's not an extension (that feels like a preprocessor for something already compatible with the flux pipeline) problem, only that I started experiencing it with kontext and I'm finding some help, I never used taesd with base flux dev before and quite possibly this would recur with base flux dev too

**Comment by DenOfEquity** [comment]
I have minor version differences but they shouldn't be relevant: transformers 4.51.3, tokenizers 0.21.1, accelerate 0.33.0, safetensors 0.5.3, huggingface_hub 0.30.0
Updating transformers can break Florence-2 Space but this version seems OK.
If you want to be ultra-sure, revert to `requirements_versions.txt` from the Forge repo.

Using xformers?

>Switching to full vae encoding /decoding the previews are fine but only the final image is black

I use TAE for previews, and I've tested it with encoding and final decode without issue. Still worth trying deleting the models and letting Forge redownload.

**Comment by realisticdreamer114514** [comment]
I already tried letting the taesd models redownload but it didnt fix anything
and I use xformers==0.0.29.post3

**Comment by realisticdreamer114514** [comment]
would torch 2.5.1 + xformers 0.0.29.post1 be more stable than 2.6.0 + 0.0.29.post3 (current venv) for 20s series cards? im thinking if i should downgrade for testing tho its not that likely the culprit

**Comment by realisticdreamer114514** [comment]
`--disable-xformers` and `--all-in-fp16` doesnt help (`--all-in-fp32` will oom even if i enable never oom)
interestingly enough the `"WTF! VAE"` warning you told me to insert isnt triggered with my black images, which might be something to pinpoint with

**Comment by DenOfEquity** [comment]
Move the NaNs check to the end of `patched_flux_forward`, immediately before the return, and check `out` instead of `k_latent`.
Are you still getting failures mid-generation, or just at end? And are previews looking correct, not just showing *something*?

**Comment by realisticdreamer114514** [comment]
> Move the NaNs check to the end of patched_flux_forward, immediately before the return, and check out 

ok, got the `"WTF! VAE"`  warning showing up every step from step 4 on. before that previews r normal (as in something is getting edited or not based on how well-written my prompts r). what kind of unusual *something* do u have in mind?

**Comment by realisticdreamer114514** [comment]
ok on a closer look the previews of failures often have those random blackish spots before turning black which seem like signs of impending nans. those might be from improper taesd encoding. im switching to full encode for testing

**Comment by realisticdreamer114514** [comment]
@DenOfEquity can you test again with **taesd encode** and these input images + settings (no need to get the particular clip l & ae I'm using)
`add the grey hat in the other image on top of the man's head while preserving the man's features and outfit <lora:FLUX.1-Turbo-Alpha-8steps:1>
Steps: 8, Sampler: Euler, Schedule type: Simple, CFG scale: 1, Distilled CFG Scale: 3, Seed: 1298775907, Size: 1024x1024, Model: flux1-kontext-dev_bnb-nf4, Token merging ratio: 0.2, bc_enabled: True, bc_method: TeaCache, bc_threshold: 0.25, bc_nocache_steps: 3, bc_skip_limit: 1, bc_always_last: False, SGM noise multiplier: True, Sigma noise: 1.03, VAE Decoder: TAESD, Diffusion in Low Bits: bnb-nf4 (fp16 LoRA), Module 1: flux_ViT-L-14-TEXT-detail-improved-hiT-GmP-TE-only-HF, Module 2: t5-v1_1-xxl-encoder-Q4_K_S, Module 3: fluxVAEW_w2`

<img width="1228" height="1228" alt="Image" src="https://github.com/user-attachments/assets/c1819a12-0236-46a9-b412-921e935d93ea" />

![Image](https://github.com/user-attachments/assets/30553f36-5bdc-4b7b-9efa-13fb6c5333ca)

preview on my device/setup, notice the grey/black spots in previews and how the nans happen during diffusion rather than tae decoding:

https://github.com/user-attachments/assets/1376a27b-05c6-409e-a7a6-07580e704aba

(btw how should I do proper latent stitching with 2 input images so that obj/details/style of pic 2 is transferred into pic 1? intuitively its drag & dropping 2 ref pics into your ext but most of the time pic 2 doesnt seem to be processed at all

this is what I tried to do with the current eg but as u see in the preview pic 2 is ignored

<img width="885" height="555" alt="Image" src="https://github.com/user-attachments/assets/bd137454-3e9a-4935-9cb9-fb43f1c75ac3" />

prompt issue? input images issue? ext issue? something not planned to be supported? too dumb to get this)

**Comment by realisticdreamer114514** [comment]
after the `devices.dtype_vae` patch, the issue is gone for fp32 taef1 but at the cost of slow (almost as slow as fp32 vae) diffusion, this issue is specifically taef1 encoding in fp16
